module/mailnews/mailread.c

```diff
@@ -353,3 +395,153 @@ void UserMailReply(int msg_num, Bool reply_all)
+      currentSortColumn = nSortColumn;
+      currentSortAscending = true;
+   }
+
+   LPARAM lParamSort = (currentSortAscending ? 1 : -1) * (currentSortColumn + 1);
+
+   if (hbmUpArrow != NULL && hbmDownArrow != NULL)
+   {
+      ListView_SetHeaderSortImage(pLVInfo->hdr.hwndFrom, currentSortColumn, currentSortAscending);
+   }
+
+   // Sort list
+   ListView_SortItemsEx(pLVInfo->hdr.hwndFrom, CompareMailListItems, lParamSort);
+}
+
+/****************************************************************************/
+/*
+ * CompareMailListItems: Comparison function for sorting items in the dialog
+ * list view.
+ */
+int CALLBACK CompareMailListItems(LPARAM lParam1, LPARAM lParam2, LPARAM lParamSort)
+{
+   bool bSortAscending = (lParamSort > 0);
+   int nColumn = abs(lParamSort) - 1;
+   char szBuf1[MAX_SUBJECT], szBuf2[MAX_SUBJECT];
+   int time1 = 0;
+   int time2 = 0;
+
+   HWND hListView = GetDlgItem(hReadMailDlg, IDC_MAILLIST);
+
+   // Extract the msg_num values from the list view items
+   ListView_GetItemText(hListView, lParam1, COL_ORDER, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, COL_ORDER, szBuf2, sizeof(szBuf2));
+   int msg_num1 = atoi(szBuf1);
+   int msg_num2 = atoi(szBuf2);
+
+   ListView_GetItemText(hListView, lParam1, nColumn, szBuf1, sizeof(szBuf1));
+   ListView_GetItemText(hListView, lParam2, nColumn, szBuf2, sizeof(szBuf2));
+
+   switch (nColumn)
+   {
+   case COL_ORDER:
+      return bSortAscending ? atoi(szBuf1) - atoi(szBuf2) : atoi(szBuf2) - atoi(szBuf1);
+   case COL_FROM:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_SUBJECT:
+      return bSortAscending ? stricmp(szBuf1, szBuf2) : stricmp(szBuf2, szBuf1);
+   case COL_TIME:
+      time1 = mail_date_map[msg_num1];
+      time2 = mail_date_map[msg_num2];
+
+      return bSortAscending ? (time1 - time2) : (time2 - time1);
+   default:
+      debug(("Unhandled column (column #%d) in CompareMailListItems\n", nColumn));
+      return 0;
+   }
+}
+
+/***************************************************************************/
+/*
+ * StringToTimestamp converts mail date strings to integers for comparison
+ */
+/***************************************************************************/
+int StringToTimestamp(const char *dateStr)
+{
+   struct tm tm = {0};
+   tm.tm_isdst = -1; // Handle daylight saving time
+
+   // Find the first space character after the day of the week
+   const char *remainingStr = std::strchr(dateStr, ' ');
+   if (remainingStr == nullptr)
+   {
+      debug(("Failed to find space after day of the week: %s\n", dateStr));
+      return 0;
+   }
+
+   // Skip past the space character
+   remainingStr++;
+
+   // Use a string stream to parse the date and time, leveraging locale
+   // to support the translation of month strings.
+   std::istringstream input(remainingStr);
+   input.imbue(std::locale(localeName));
+   input >> std::get_time(&tm, "%b %d, %Y %H:%M");
```

Comment:
Is it possible to use %a in get_time rather than manually skipping the day of the week in the string?

https://en.cppreference.com/w/cpp/io/manip/get_time

Reply:
Oops! I thought it would contribute to the timestamp generation but that's not the case.

> Renormalizes local calendar time expressed as a struct tm object and also converts it to time since epoch as a time_t object. **arg->tm_wday** and arg->tm_yday are ignored.